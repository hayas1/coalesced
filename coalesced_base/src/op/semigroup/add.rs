use coalesced_derive::ConstructionUse;

use crate::{op::Construction, reverse::Reversed, semigroup::Semigroup};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Hash, ConstructionUse)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[construction(op = Add)]
pub struct Added<T: std::ops::Add<Output = T>>(pub T);
impl<T: std::ops::Add<Output = T>> Semigroup for Added<T> {
    fn semigroup_op(base: Self, other: Self) -> Self {
        Added(base.0 + other.0)
    }
}

#[cfg(test)]
mod tests {
    use crate::semigroup::tests::{assert_associative_law, assert_reversed_associative_law};

    use super::*;

    #[test]
    fn test_add_as_semigroup_op() {
        let (a, b, c) = (Added(1), Added(2), Added(3));
        assert_associative_law(a, b, c);
        assert_reversed_associative_law(a, b, c);
    }

    #[test]
    fn test_add() {
        let (a, b) = (Added(1), Added(2));
        assert_eq!(a.add(b).into_inner(), 3);

        let (ra, rb) = (Reversed(a), Reversed(b));
        assert_eq!(ra.add(rb).0.into_inner(), 3);
    }
}
