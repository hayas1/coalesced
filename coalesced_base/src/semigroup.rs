use crate::annotate::{Annotate, Annotated};

pub trait Semigroup {
    fn semigroup_op(base: Self, other: Self) -> Self;
}
pub trait AnnotatedSemigroup<A>: Sized {
    fn annotated_op(base: Annotated<Self, A>, other: Annotated<Self, A>) -> Annotated<Self, A>;
    fn default_semigroup_op(
        base: Self,
        other: Self,
        base_annotation: A,
        other_annotation: A,
    ) -> Self {
        Self::annotated_op(
            base.annotated(base_annotation),
            other.annotated(other_annotation),
        )
        .value
    }
}
impl<T: AnnotatedSemigroup<A>, A> Semigroup for Annotated<T, A> {
    fn semigroup_op(base: Self, other: Self) -> Self {
        AnnotatedSemigroup::annotated_op(base, other)
    }
}

#[cfg(any(test, feature = "test"))]
pub mod tests {
    use std::fmt::Debug;

    use crate::op::reverse::Reversed;

    use super::*;

    pub fn associative_law<T: Semigroup + Clone>(a: T, b: T, c: T) -> (T, T) {
        let ab_c = T::semigroup_op(T::semigroup_op(a.clone(), b.clone()), c.clone());
        let a_bc = T::semigroup_op(a.clone(), T::semigroup_op(b.clone(), c.clone()));
        (ab_c, a_bc)
    }
    pub fn assert_associative_law<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        let (ab_c, a_bc) = associative_law(a, b, c);
        assert_eq!(ab_c, a_bc);
    }

    pub fn reverse<T: Semigroup + Clone>(a: T, b: T) -> (T, T) {
        let (ra, rb) = (Reversed(a.clone()), Reversed(b.clone()));
        (T::semigroup_op(a, b), Reversed::<T>::semigroup_op(rb, ra).0)
    }
    pub fn assert_reversed_associative_law<T: Semigroup + Clone + PartialEq + Debug>(
        a: T,
        b: T,
        c: T,
    ) {
        let (ra, rb, rc) = (Reversed(a), Reversed(b), Reversed(c));
        assert_associative_law(ra, rb, rc);
    }
}
