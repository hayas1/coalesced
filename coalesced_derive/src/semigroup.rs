use proc_macro2::TokenStream;
use syn::DeriveInput;

use crate::{constant::ConstantExt, semigroup::attr::ContainerAttr};

mod attr;
mod implementor;

pub fn gen_semigroup<C: ConstantExt>(derive: &DeriveInput) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let attr = ContainerAttr::new(derive)?;

    todo!()
}
