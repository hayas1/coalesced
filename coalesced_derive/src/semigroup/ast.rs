use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::{
    parse_quote, Data, DataEnum, DataStruct, DataUnion, DeriveInput, Field, FieldValue, Fields,
    FieldsNamed, Ident, ItemImpl,
};

use crate::{
    constant::Constant,
    error::SemigroupError,
    semigroup::attr::{ContainerAttr, FieldAttr},
};

#[derive(Debug, Clone)]
pub struct Semigroup<'a> {
    derive: &'a DeriveInput,
    semigroup_trait: SemigroupTrait<'a>,
}
impl ToTokens for Semigroup<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Self {
            semigroup_trait, ..
        } = self;
        tokens.extend(quote::quote! {
            #semigroup_trait
        });
    }
}
impl<'a> Semigroup<'a> {
    pub fn new(
        constant: &'a Constant,
        derive: &'a DeriveInput,
        attr: &'a ContainerAttr,
    ) -> syn::Result<Self> {
        match &derive.data {
            Data::Enum(DataEnum { enum_token, .. }) => Err(syn::Error::new_spanned(
                enum_token,
                SemigroupError::UnsupportedEnum,
            )),
            Data::Struct(data_struct) => {
                let semigroup_trait = SemigroupTrait::new(constant, derive, attr, data_struct)?;
                Ok(Self {
                    derive,
                    semigroup_trait,
                })
            }
            Data::Union(DataUnion { union_token, .. }) => Err(syn::Error::new_spanned(
                union_token,
                SemigroupError::UnsupportedUnion,
            )),
        }
    }
}

#[derive(Debug, Clone)]
pub struct SemigroupTrait<'a> {
    constant: &'a Constant,
    derive: &'a DeriveInput,
    attr: &'a ContainerAttr,
    data_struct: &'a DataStruct,
}
impl ToTokens for SemigroupTrait<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let impl_semigroup = self
            .impl_semigroup()
            .as_ref()
            .map(ToTokens::to_token_stream)
            .unwrap_or_else(syn::Error::to_compile_error);
        impl_semigroup.to_tokens(tokens)
    }
}
impl<'a> SemigroupTrait<'a> {
    pub fn new(
        constant: &'a Constant,
        derive: &'a DeriveInput,
        attr: &'a ContainerAttr,
        data_struct: &'a DataStruct,
    ) -> syn::Result<Self> {
        Ok(Self {
            constant,
            derive,
            data_struct,
            attr,
        })
    }
    pub fn impl_semigroup(&self) -> syn::Result<ItemImpl> {
        let Self {
            constant,
            derive,
            attr,
            ..
        } = self;
        match &self.data_struct.fields {
            Fields::Named(fields_named) => {
                let named = SemigroupTraitNamed::new(constant, derive, attr, fields_named);
                named.impl_semigroup()
            }
            Fields::Unnamed(fields_unnamed) => todo!(),
            Fields::Unit => todo!(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct SemigroupTraitNamed<'a> {
    constant: &'a Constant,
    derive: &'a DeriveInput,
    attr: &'a ContainerAttr,
    fields_named: &'a FieldsNamed,
}
impl ToTokens for SemigroupTraitNamed<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let impl_semigroup = self
            .impl_semigroup()
            .as_ref()
            .map(ToTokens::to_token_stream)
            .unwrap_or_else(syn::Error::to_compile_error);
        impl_semigroup.to_tokens(tokens)
    }
}
impl<'a> SemigroupTraitNamed<'a> {
    pub fn new(
        constant: &'a Constant,
        derive: &'a DeriveInput,
        attr: &'a ContainerAttr,
        fields_named: &'a FieldsNamed,
    ) -> Self {
        Self {
            constant,
            derive,
            attr,
            fields_named,
        }
    }

    pub fn impl_semigroup(&self) -> syn::Result<ItemImpl> {
        let Self {
            constant,
            derive,
            attr,
            fields_named,
            ..
        } = self;
        let Constant {
            path_semigroup,
            ident_semigroup_op,
            ..
        } = constant;
        let DeriveInput { ident, .. } = derive;
        let fields = SemigroupTraitField::new_fields(constant, derive, attr, &fields_named.named)?;
        Ok(parse_quote! {
            impl #path_semigroup for #ident {
                fn #ident_semigroup_op(base: Self, other: Self) -> Self {
                    Self {
                        #(#fields),*
                    }
                }
            }
        })
    }
}

pub struct SemigroupTraitField<'a> {
    constant: &'a Constant,
    derive: &'a DeriveInput,
    container_attr: &'a ContainerAttr,
    index: usize,
    field: &'a Field,
    field_attr: FieldAttr,
}
impl ToTokens for SemigroupTraitField<'_> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let impl_field = self.impl_semigroup_field();
        impl_field.to_tokens(tokens)
    }
}
impl<'a> SemigroupTraitField<'a> {
    pub fn new(
        constant: &'a Constant,
        derive: &'a DeriveInput,
        container_attr: &'a ContainerAttr,
        index: usize,
        field: &'a Field,
    ) -> syn::Result<Self> {
        let field_attr = FieldAttr::new(field)?;
        Ok(Self {
            constant,
            derive,
            container_attr,
            index,
            field,
            field_attr,
        })
    }
    pub fn new_fields(
        constant: &'a Constant,
        derive: &'a DeriveInput,
        container_attr: &'a ContainerAttr,
        fields: impl 'a + IntoIterator<Item = &'a Field>,
    ) -> syn::Result<Vec<Self>> {
        fields
            .into_iter()
            .enumerate()
            .map(move |(index, field)| Self::new(constant, derive, container_attr, index, field))
            .collect()
    }
    pub fn impl_semigroup_field(&self) -> FieldValue {
        match &self.field.ident {
            Some(ident) => self.impl_semigroup_field_named(ident),
            None => todo!(),
        }
    }

    pub fn impl_semigroup_field_named(&self, ident: &Ident) -> FieldValue {
        let Self {
            constant:
                Constant {
                    path_semigroup,
                    ident_semigroup_op,
                    path_construction_trait,
                    ..
                },
            container_attr,
            field_attr,
            ..
        } = self;
        field_attr
            .with(container_attr)
            .map(|path| {
                parse_quote! {
                    #ident: <#path<_> as #path_construction_trait<_>>::lift_op(base.#ident, other.#ident)
                }
            })
            .unwrap_or_else(|| {
                parse_quote! {
                    #ident: #path_semigroup::#ident_semigroup_op(base.#ident, other.#ident)
                }
            })
    }
}
