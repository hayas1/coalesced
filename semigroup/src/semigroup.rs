use crate::Annotated;

/// [`Semigroup`] represents a binary operation that satisfies the following properties
/// 1. *Closure*: `op: T × T → T`
/// 2. *Associativity*: `op(op(a, b), c) = op(a, op(b, c))`
///
/// # Examples
/// ## Deriving
/// When fields do not implement [`Semigroup`], use `with` attribute.
/// ```
/// use semigroup::Semigroup;
/// #[derive(Debug, Clone, PartialEq, Semigroup)]
/// #[semigroup(with = "semigroup::op::Coalesce")]
/// pub struct ExampleStruct<'a> {
///     pub str: Option<&'a str>,
///     #[semigroup(with = "semigroup::op::Overwrite")]
///     pub boolean: bool,
///     #[semigroup(with = "semigroup::op::Sum")]
///     pub sum: u32,
/// }
///
/// let a = ExampleStruct { str: None, boolean: true, sum: 1 };
/// let b = ExampleStruct { str: Some("ten"), boolean: false, sum: 10 };
/// let c = ExampleStruct { str: None, boolean: false, sum: 100 };
///
/// // #[test]
/// semigroup::assert_semigroup!(&a, &b, &c);
/// assert_eq!(a.semigroup(b).semigroup(c), ExampleStruct { str: Some("ten"), boolean: false, sum: 111 });
/// ```
///
/// ## Construction
/// [`Semigroup`] can be constructed by [`crate::Construction`].
///
/// Some operations are already provided by [`crate::op`].
/// ```
/// use semigroup::{Construction, Semigroup};
///
/// #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Default, Hash, Construction)]
/// pub struct Sum(u64);
/// impl Semigroup for Sum {
///     fn op(base: Self, other: Self) -> Self {
///         Self(base.0 + other.0)
///     }
/// }
///
/// let (a, b, c) = (Sum(1), Sum(2), Sum(3));
/// // #[test]
/// semigroup::assert_semigroup!(&a, &b, &c);
/// assert_eq!(a.semigroup(b).semigroup(c), Sum(6));
/// ```
///
/// # Testing
/// Use [`crate::assert_semigroup!`] macro.
///
/// The *closure* property is guaranteed by Rust’s type system,
/// but *associativity* must be verified manually using [`crate::assert_semigroup!`].
pub trait Semigroup {
    fn op(base: Self, other: Self) -> Self;
    fn semigroup(self, other: Self) -> Self
    where
        Self: Sized,
    {
        Semigroup::op(self, other)
    }
}

/// [`AnnotatedSemigroup`] is a [`Semigroup`] that has an annotation, such as [`crate::Annotate`].
pub trait AnnotatedSemigroup<A>: Sized + Semigroup {
    fn annotated_op(base: Annotated<Self, A>, other: Annotated<Self, A>) -> Annotated<Self, A>;
}

#[cfg(feature = "test")]
pub mod test_semigroup {
    use std::fmt::Debug;

    use rand::seq::IndexedRandom;

    use crate::{
        combine::test_combine::{assert_combine_iter, assert_lazy},
        commutative::test_commutative::{assert_reverse_associative_law, assert_reverse_reverse},
    };

    use super::*;

    /// Assert that the given type satisfies the *semigroup* property.
    ///
    /// # Usage
    /// - 1 argument: iterator of more than 3 items that implements [`Semigroup`].
    /// - More than 3 arguments: items that implements [`Semigroup`].
    ///
    /// # Examples
    /// ```
    /// use semigroup::{assert_semigroup, op::Coalesce};
    ///
    /// let a = Coalesce(Some(1));
    /// let b = Coalesce(None);
    /// let c = Coalesce(Some(3));
    /// assert_semigroup!(a, b, c);
    ///
    /// let v = vec![a, b, c];
    /// assert_semigroup!(&v);
    /// ```
    ///
    /// # Panics
    /// - If the given function does not satisfy the *semigroup* property.
    /// - The input iterator has less than 3 items.
    ///
    /// ```compile_fail
    /// use semigroup::{assert_semigroup, op::Coalesce};
    /// let a = Coalesce(Some(1));
    /// let b = Coalesce(None);
    /// assert_semigroup!(a, b);
    /// ```
    #[macro_export]
    macro_rules! assert_semigroup {
        ($a:expr, $b: expr, $($tail: expr),*) => {
            {
                let v = vec![$a, $b, $($tail),*];
                $crate::assert_semigroup!(&v)
            }
        };
        ($v:expr) => {
            {
                let (a, b, c) = $crate::test_semigroup::pick3($v);
                $crate::test_semigroup::assert_semigroup_impl(a.clone(), b.clone(), c.clone());
            }
        };
    }

    pub fn pick3<T: Clone>(data: &[T]) -> (T, T, T) {
        data.choose_multiple_array(&mut rand::rng())
            .map(|[a, b, c]| (a, b, c))
            .expect("failed to pick 3 items")
    }

    pub fn assert_semigroup_impl<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        assert_associative_law(a.clone(), b.clone(), c.clone());
        assert_reverse_reverse(a.clone(), b.clone(), c.clone());
        assert_reverse_associative_law(a.clone(), b.clone(), c.clone());
        assert_combine_iter(a.clone(), b.clone(), c.clone());
        assert_lazy(a.clone(), b.clone(), c.clone());
        #[cfg(feature = "monoid")]
        crate::test_monoid::assert_option_monoid(a.clone(), b.clone(), c.clone());
    }

    pub fn assert_associative_law<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        let ab_c = Semigroup::op(Semigroup::op(a.clone(), b.clone()), c.clone());
        let a_bc = Semigroup::op(a.clone(), Semigroup::op(b.clone(), c.clone()));
        assert_eq!(ab_c, a_bc);
    }
}
