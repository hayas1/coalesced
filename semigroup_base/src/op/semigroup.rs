#[cfg(feature = "monoid")]
pub mod gcd;
#[cfg(feature = "monoid")]
pub mod lcm;
pub mod max;
pub mod min;
pub mod prod;
pub mod sum;
pub mod xor;
