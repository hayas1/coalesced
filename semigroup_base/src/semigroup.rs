use crate::annotate::Annotated;

pub trait Semigroup {
    fn semigroup_op(base: Self, other: Self) -> Self;
    fn semigroup(self, other: Self) -> Self
    where
        Self: Sized,
    {
        Semigroup::semigroup_op(self, other)
    }
}
pub trait AnnotatedSemigroup<A>: Sized + Semigroup {
    fn annotated_op(base: Annotated<Self, A>, other: Annotated<Self, A>) -> Annotated<Self, A>;
}

#[cfg(any(test, feature = "test"))]
pub mod tests {
    use std::fmt::Debug;

    use crate::{
        iter::tests::assert_lazy_evaluation_iter,
        reverse::tests::{assert_reverse, assert_reverse_associative_law},
    };

    use super::*;

    #[macro_export]
    macro_rules! assert_semigroup_op {
        ($a:expr, $b: expr, $c: expr) => {
            $crate::semigroup::tests::assert_semigroup_op_impl($a.clone(), $b.clone(), $c.clone());
            $crate::monoid::tests::assert_option_monoid($a.clone(), $b.clone(), $c.clone());
        };
    }
    pub use assert_semigroup_op;

    pub fn assert_semigroup_op_impl<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        assert_associative_law(a.clone(), b.clone(), c.clone());
        assert_reverse(a.clone(), b.clone(), c.clone());
        assert_reverse_associative_law(a.clone(), b.clone(), c.clone());
        assert_lazy_evaluation_iter(a.clone(), b.clone(), c.clone());
    }

    pub fn assert_associative_law<T: Semigroup + Clone + PartialEq + Debug>(a: T, b: T, c: T) {
        let ab_c = T::semigroup_op(T::semigroup_op(a.clone(), b.clone()), c.clone());
        let a_bc = T::semigroup_op(a.clone(), T::semigroup_op(b.clone(), c.clone()));
        assert_eq!(ab_c, a_bc);
    }
}
