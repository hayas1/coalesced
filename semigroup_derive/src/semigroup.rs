use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::DeriveInput;

use crate::{
    constant::ConstantExt,
    semigroup::{ast::Semigroup, attr::ContainerAttr},
};

mod ast;
mod attr;

pub fn impl_semigroup<C: ConstantExt>(derive: &DeriveInput) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let attr = ContainerAttr::new(derive)?;
    let semigroup = Semigroup::new(&constant, derive, &attr)?;
    Ok(semigroup.into_token_stream())
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use crate::constant::{External, Internal};

    use super::*;

    #[rstest]
    #[case::semigroup_annotated(
        "semigroup_annotated",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(annotated)]
            pub struct NamedStruct {
                #[semigroup(with = "semigroup::op::Overwrite")]
                pub foo: String,
                pub bar: Option<u32>,
                pub baz: semigroup::op::Overwrite<bool>,
            }
        },
    )]
    #[case::semigroup_not_annotated(
        "semigroup_not_annotated",
        impl_semigroup::<Internal>,
        syn::parse_quote! {
            #[derive(SemigroupPriv)]
            #[semigroup(with = "semigroup::op::Overwrite")]
            pub struct UnnamedStruct<T: std::ops::Add> (
                #[semigroup(with = "semigroup::op::Added")]
                T,
                u64
            );
        },
    )]
    #[case::semigroup_custom_annotation(
        "semigroup_custom_annotation",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(annotated, annotation_param = X, with = "semigroup::op::Overwrite")]
            pub struct NamedStruct{
                pub foo: String,
                pub bar: Option<u32>,
                pub baz: bool,
            }
        },
    )]
    #[case::semigroup_monoid(
        "semigroup_monoid",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(monoid, commutative)]
            pub struct MonoidStruct{
                #[semigroup(with = "semigroup::op::Sum")]
                pub sum: u32,
            }
        },
    )]
    fn test_derive_semigroup_snapshot(
        #[case] case: &str,
        #[case] f: impl Fn(&DeriveInput) -> syn::Result<TokenStream>,
        #[case] input: DeriveInput,
    ) {
        let generated = f(&input).unwrap();
        let formatted = prettyplease::unparse(&syn::parse2(generated).unwrap());
        insta::with_settings!({ snapshot_path => "../tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_snapshot!(case, formatted);
        });
    }
}
