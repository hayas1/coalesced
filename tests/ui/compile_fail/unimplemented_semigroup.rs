use semigroup::Semigroup;

#[derive(Debug, Clone, PartialEq, Semigroup)]
pub struct NamedStruct<'a> {
    pub num: Option<u32>,
    #[semigroup(with = "semigroup::op::coalesce::Coalesce")]
    pub str: Option<&'a str>,
}

fn main() {
    let a = NamedStruct {
        num: Some(1),
        str: None,
    };
    let b = NamedStruct {
        num: None,
        str: Some("ten"),
    };

    let c = a.semigroup(b);

    assert_eq!(
        c,
        NamedStruct {
            num: Some(1),
            str: Some("ten"),
        }
    );
}
