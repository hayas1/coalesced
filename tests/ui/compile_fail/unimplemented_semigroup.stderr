error[E0277]: the trait bound `Option<u32>: Semigroup` is not satisfied
 --> tests/ui/compile_fail/unimplemented_semigroup.rs:3:35
  |
3 | #[derive(Debug, Clone, PartialEq, Semigroup)]
  |                                   ^^^^^^^^^ the trait `Semigroup` is not implemented for `Option<u32>`
  |
  = help: the following other types implement trait `Semigroup`:
            Annotated<T, A>
            Coalesce<T>
            NamedStruct<'a>
            Replace<T>
            coalesced::Reverse<T>
            coalesced::op::annotation::concat::Concat<T>
            coalesced::op::semigroup::add::Add<T>
  = note: this error originates in the derive macro `Semigroup` (in Nightly builds, run with -Z macro-backtrace for more info)
